File: js101/lesson_2/mortgage-calculator.js
/* Solution to assignment 19 of lesson 2: Mortgage Calculator

Inputs (from user):
- Loan amount
- Annual Percentage Rate (APR)
- Loan duration in years

Calculation:
- Monthly interest rate
- Loan duration in months

Output:
- Monthly payment

Formula:
let discountFactor = Math.pow((1 + monthlyInterestRate), -loanDurationInMonths);

let monthlyPayment = loanAmount * (monthlyInterestRate / (1 - discountFactor));

Pseudocode:
Pseudocode ask the user for another calculation:
WHILE iterator equals true
  GET loanAmount from user
    ("What is the loan amount?")
  GET annualInterestRate from user
    ("What is the annual interest rate? (Example: 5 for 5% or 2.5 for 2.5%)")
  GET loanDurationInYears from user
    ("What is the loan duration (in years)?")
  PRINT monthlyPayment
  GET input from user (continue? y/n)

SET loanDurationInMonths = loanDurationInYears / 12
SET monthlyInterestRate = annualInterestRate / 12
SET discountFactor = Math.pow((1 + monthlyInterestRate), -loanDurationInMonths)
SET monthlyPayment = loanAmount * (monthlyInterestRate / (1 - discountFactor))

PRINT loanDurationInMonths
PRINT monthlyInterestRate
PRINT monthlyPayment
*/

const readline = require('readline-sync');

// Prettier console log when asking questions
function prompt(message) {
  console.log(`=> ${message}`);
}

// Check if inputted number is invalid
function invalidNumber(num) {
  return num.trimStart() === '' || Number.isNaN(Number(num));
}

// Prompt user for number
function getNumber(promptMessage) {
  prompt(promptMessage);

  let num = readline.question();

  while (invalidNumber(num)) {
    prompt("Hmm... that doesn't look like a valid number.");
    num = readline.question();
  }

  return Number(num);
}

function calculateMonthlyPayment(loanAmount,
  annualInterestRate, loanDurationInYears) {

  // Calculate monthly rates
  let loanDurationInMonths = loanDurationInYears * 12;
  let monthlyInterestRate = annualInterestRate / 100 / 12;
  let monthlyInterestRatePercent = monthlyInterestRate * 100;

  // Calculate monthly dollar payment
  let discountFactor = Math.pow((1 + monthlyInterestRate),
    -loanDurationInMonths);
  let monthlyPayment = loanAmount * (monthlyInterestRate /
    (1 - discountFactor));

  console.log(`Loan duration: ${loanDurationInMonths} months`);
  console.log(`Monthly interest rate: ${monthlyInterestRatePercent.toFixed(2)}%`);
  console.log(`Monthly payment: $${monthlyPayment.toFixed(2)}`);
}

// Ask the user for the numbers and operation
while (true) {
  const loanAmount = getNumber("What is the loan amount?");
  const annualInterestRate = getNumber("What is the annual interest rate? (Example: 5 for 5% or 2.5 for 2.5%)");
  const loanDurationInYears = getNumber("What is the loan duration (in years)?");

  console.log(calculateMonthlyPayment(
    loanAmount, annualInterestRate, loanDurationInYears));

  if (readline.question("Do another calculation? (y/n) ").toLowerCase() === 'n') break;
}

// Test Case 1:
// Loan Amount: $150,000
// Annual Interest Rate: 4.5%
// Loan Duration: 30 years
// calculateMonthlyPayment(150000, 4.5, 30);
// Expected Output:
// Loan duration: 360 months
// Monthly interest rate: 0.38%
// Monthly payment: $760.03

// Test Case 2:
// Loan Amount: $350,000
// Annual Interest Rate: 3.92%
// Loan Duration: 20 years
// calculateMonthlyPayment(350000, 3.92, 20);
// Expected Output:
// Loan duration: 240 months
// Monthly interest rate: 0.33%
// Monthly payment: $2071.84

// Test Case 3:
// Loan Amount: $500,000
// Annual Interest Rate: 2.75%
// Loan Duration: 15 years
// calculateMonthlyPayment(500000, 2.75, 15);
// Expected Output:
// Loan duration: 180 months
// Monthly interest rate: 0.23%
// Monthly payment: $3381.29

// Test Case 4:
// Loan Amount: $250,000
// Annual Interest Rate: 3.25%
// Loan Duration: 10 years
// calculateMonthlyPayment(250000, 3.25, 10);
// Expected Output:
// Loan duration: 120 months
// Monthly interest rate: 0.27%
// Monthly payment: $2432.35

File: python-track/py101/lesson_2/mortgage_calculator.py
# Build a mortgage calculator that determines the monthly payment, 
# assuming that interest is compounded monthly.

# Inputs:
# X Loan amount
# X Loan duration in years
# X Yearly interest rate

# Calculate:
# X Monthly interest rate (APR / 12)
# X Loan duration in months

# Output:
# X Print monthly payment amount as dollars and cents
#   (e.g., $123.45 or $371.00.)

# Questions:
# X The input loan duration is in years?
#   X Let's assume, YES
# X What input is accepted?
#   X Loan amount: float (10000.5)
#   X Loan duration in years: float (2.5)
#   X APR: float (5.5)

# Formula:
# Monthly payment = loan amount * (month interest rate /
#   (1 - (1 + monthly interest rate) ** (-loan duration in months)))

def prompt(message):
    print(f"==> {message}")

def invalid_number(input):
    try:
        number = float(input)
        if number <= 0:
            raise ValueError(f"Value must be > 0: {input}")
    except ValueError:
        return True
    
    return False


prompt("Welcome to Mortgage Calculator!")

while True:
    # Ask user for loan amount
    prompt("How much do you want to loan?")
    loan_amount_input = input()

    while invalid_number(loan_amount_input):
        prompt("That's not a valid number, please try again.")
        loan_amount_input = input()

    loan_amount = float(loan_amount_input)

    # Ask user for loan duration
    prompt("How long should the loan be (in years)?")
    loan_duration_years_input = input()

    while invalid_number(loan_duration_years_input):
        prompt("That's not a valid number, please try again.")
        loan_duration_years_input = input()

    loan_duration_years = float(loan_duration_years_input)

    # Ask user for yearly interest rate
    prompt("What is the yearly interest rate (APR)?")
    prompt(f"Example: 5 for 5% or 2.5 for 2.5%")
    yearly_interest_rate_input = input()

    while invalid_number(yearly_interest_rate_input):
        prompt("That's not a valid number, please try again.")
        yearly_interest_rate_input = input()

    yearly_interest_rate = float(yearly_interest_rate_input)
            
    # Calculate monthly interest rate and loan duration in months
    monthly_interest_rate = (yearly_interest_rate / 12) / 100
    loan_duration_months = loan_duration_years * 12

    # Calculate monthly payment
    monthly_payment = loan_amount * (monthly_interest_rate / (
        1 - (1 + monthly_interest_rate) ** (-loan_duration_months)
    ))

    # Show monthly payment amount
    prompt(f"Your monthly payment is: ${round(monthly_payment, 2)}")
    print("--------------------------------------")
    prompt("Do you want to do another calculation? (y/n)")
    continue_calculating = input().strip().lower()

    if continue_calculating == '':
        break
    if continue_calculating[0] != 'y':
        break
